internal/crypto/cipher.go

```diff
@@ -0,0 +1,208 @@
+package crypto
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"fmt"
+	"hash"
+
+	"github.com/optable/match/internal/util"
+	"github.com/zeebo/blake3"
+)
+
+/*
+Various cipher suite implementation in golang
+*/
+
+const (
+	GCM = iota
+	XORBlake3
+
+	nonceSize = 12 //aesgcm NonceSize
+)
+
+// PseudorandomCode is implemented as follows:
+// C(x) = AES(x[:16]) || AES(x[16:32]) || AES(x[32:48]) || AES(x[48:])
+// secretKey is a 16 byte slice for AES-128
+// on success, pseudorandomCode returns a byte slice of 64 bytes.
+func PseudorandomCode(aesBlock cipher.Block, src []byte) (dst []byte) {
+	dst = make([]byte, aes.BlockSize*4)
+	// effectively pad src
+	copy(dst, src)
+
+	// encrypt
+	aesBlock.Encrypt(dst[:aes.BlockSize], dst[:aes.BlockSize])
+	if len(src) <= aes.BlockSize {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize:aes.BlockSize*2], dst[aes.BlockSize:aes.BlockSize*2])
+	if len(src) <= aes.BlockSize*2 {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize*2:aes.BlockSize*3], dst[aes.BlockSize*2:aes.BlockSize*3])
+	if len(src) <= aes.BlockSize*3 {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize*3:], dst[aes.BlockSize*3:])
+	return dst
+}
+
+// PseudorandomCodeWithHashIndex is implemented as follows:
+// C(x) = AES(x[:16]) || AES(x[16:32]) || AES(x[32:48]) || AES(x[48:])
+// secretKey is a 16 byte slice for AES-128
+// on success, pseudorandomCode returns a byte slice of 64 bytes.
+func PseudorandomCodeWithHashIndex(aesBlock cipher.Block, src []byte, hIdx byte) (dst []byte) {
+	dst = make([]byte, aes.BlockSize*4)
+	// effectively pad src
+	copy(dst, src)
+	if len(src) < aes.BlockSize*4 {
+		dst[len(src)] = hIdx
+	}
+
+	// encrypt
+	aesBlock.Encrypt(dst[:aes.BlockSize], dst[:aes.BlockSize])
+	if len(src)+1 <= aes.BlockSize {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize:aes.BlockSize*2], dst[aes.BlockSize:aes.BlockSize*2])
+	if len(src)+1 <= aes.BlockSize*2 {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize*2:aes.BlockSize*3], dst[aes.BlockSize*2:aes.BlockSize*3])
+	if len(src)+1 <= aes.BlockSize*3 {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize*3:], dst[aes.BlockSize*3:])
+	return dst
+}
+
+// PseudorandomCodeHmacSHA256 is implemented as follows:
+// C(x) = HmacSHA256_key(x[:32]) || HmacSHA256_key(x[32:]
+// secretKey is a 32 byte slice for Hmac-SHA256
+// on success, pseudorandomCode returns a byte slice of 64 bytes.
+func PseudorandomCodeHmacSHA256(prf hash.Hash, src []byte) (dst []byte) {
```

Comment:
In this case, src has to contain 64 bytes exactly?

Reply:
This is an old implementation that we have left in. We can remove this.

Reply:
Yes let's remove it